File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Core/Bntx/BntxFile.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using MiniToolbox.Core.Bntx;
using MiniToolbox.Core.Bntx;

namespace MiniToolbox.Core.Bntx;

/// <summary>
/// BNTX file container (read-only). Loads and provides access to texture entries.
/// </summary>
[DebuggerDisplay("BntxFile {Name}")]
public class BntxFile : IResData
{
	internal uint FileSizeToRLT;
	internal byte[] OriginalRLTChunk;

	public string Name { get; set; }

	public string PlatformTarget
	{
		get => new string(Target);
		set => Target = value.ToCharArray();
	}

	public string VersionFull => $"{VersionMajor}.{VersionMajor2}.{VersionMinor}.{VersionMinor2}";
	public uint VersionMajor { get; set; }
	public uint VersionMajor2 { get; set; }
	public uint VersionMinor { get; set; }
	public uint VersionMinor2 { get; set; }
	public ByteOrder ByteOrder { get; private set; }
	public uint Alignment { get; set; }
	public int DataAlignment => 1 << (int)Alignment;
	public uint TargetAddressSize { get; set; }
	public uint Flag { get; set; }
	public uint BlockOffset { get; set; }
	public RelocationTable RelocationTable { get; set; }
	public char[] Target { get; set; }
	public IList<Texture> Textures { get; set; }
	public ResDict TextureDict { get; set; }
	public StringTable StringTable { get; set; }

	public BntxFile()
	{
		StringTable = new StringTable();
	}

	public BntxFile(Stream stream, bool leaveOpen = false)
	{
		using BntxFileLoader bntxFileLoader = new BntxFileLoader(this, stream, leaveOpen);
		bntxFileLoader.Execute();
	}

	public BntxFile(string fileName)
	{
		using BntxFileLoader bntxFileLoader = new BntxFileLoader(this, fileName);
		bntxFileLoader.Execute();
	}

	private void SetVersionInfo(uint Version)
	{
		VersionMajor = Version >> 24;
		VersionMajor2 = (Version >> 16) & 0xFF;
		VersionMinor = (Version >> 8) & 0xFF;
		VersionMinor2 = Version & 0xFF;
	}

	void IResData.Load(BntxFileLoader loader)
	{
		StringTable = new StringTable();
		loader.CheckSignature("BNTX");
		uint num = loader.ReadUInt32();
		uint versionInfo = loader.ReadUInt32();
		SetVersionInfo(versionInfo);
		ByteOrder = loader.ReadEnum<ByteOrder>(strict: false);
		Alignment = loader.ReadByte();
		TargetAddressSize = loader.ReadByte();
		uint num2 = loader.ReadUInt32();
		Flag = loader.ReadUInt16();
		BlockOffset = loader.ReadUInt16();
		uint num3 = loader.ReadUInt32();
		uint num4 = loader.ReadUInt32();
		Target = loader.ReadChars(4);
		int textureCount = loader.ReadInt32();
		long num5 = loader.ReadInt64();
		FileSizeToRLT = num3;
		OriginalRLTChunk = loader.LoadCustom(() => loader.ReadBytes((int)(loader.BaseStream.Length - FileSizeToRLT)), num3);
		Textures = loader.LoadCustom(delegate
		{
			IList<Texture> list = new List<Texture>();
			for (int i = 0; i < textureCount; i++)
			{
				list.Add(loader.Load<Texture>());
			}
			return list;
		}, num5);
		long num6 = loader.ReadInt64();
		TextureDict = loader.LoadDict();
		Name = loader.LoadString(null, num2 - 2);
		loader.Seek(num5 + textureCount * 8, SeekOrigin.Begin);
		StringTable.Load(loader);
	}
}

File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Api/src/bntx/BntxFile.ts
import { BinaryDataReader, ByteOrder, SeekTask } from './BinaryDataReader.js';
import { RelocationTable } from './RelocationTable.js';
import { ResDict } from './ResDict.js';
import { StringTable } from './StringTable.js';
import { Texture } from './Texture.js';
import { IResData } from './types.js';

export class BntxFile implements IResData {
    public name: string = "";
    public target: string = "";
    public versionMajor: number = 0;
    public versionMajor2: number = 0;
    public versionMinor: number = 0;
    public versionMinor2: number = 0;
    public byteOrder: ByteOrder = ByteOrder.LittleEndian;
    public alignment: number = 0;
    public targetAddressSize: number = 0;
    public flag: number = 0;
    public blockOffset: number = 0;

    public relocationTable: RelocationTable = new RelocationTable();
    public textures: Texture[] = [];
    public textureDict: ResDict = new ResDict();
    public stringTable: StringTable = new StringTable();

    get platformTarget(): string {
        return this.target;
    }

    get versionFull(): string {
        return `${this.versionMajor}.${this.versionMajor2}.${this.versionMinor}.${this.versionMinor2}`;
    }

    load(reader: BinaryDataReader): void {
        const loader = new BntxLoader(this, reader);
        loader.execute();
    }

    static fromBuffer(buffer: Buffer): BntxFile {
        const file = new BntxFile();
        const reader = new BinaryDataReader(buffer);
        file.load(reader);
        return file;
    }
}

/**
 * Helper class to mirror C# BntxFileLoader logic
 */
class BntxLoader {
    private file: BntxFile;
    private reader: BinaryDataReader;

    constructor(file: BntxFile, reader: BinaryDataReader) {
        this.file = file;
        this.reader = reader;
    }

    execute(): void {
        const reader = this.reader;
        const file = this.file;

        // Signature
        const sig = reader.readStringWithLength(4);
        if (sig !== 'BNTX') {
            throw new Error(`Invalid BNTX signature: ${sig}`);
        }

        reader.readUInt32(); // Skip padding/reserved
        const versionInfo = reader.readUInt32();
        this.setVersionInfo(versionInfo);

        file.byteOrder = reader.readUInt16() as ByteOrder;
        reader.byteOrder = file.byteOrder; // Sync reader byte order

        file.alignment = reader.readByte();
        file.targetAddressSize = reader.readByte();

        const fileNameOffset = Number(reader.readUInt32());
        file.flag = reader.readUInt16();
        file.blockOffset = reader.readUInt16();

        const relocationTableOffset = reader.readUInt32();
        const fileSizeToRLT = reader.readUInt32();

        file.target = reader.readStringWithLength(4);

        const textureCount = reader.readInt32();
        const infoPtrArrayOffset = Number(reader.readInt64());
        const dataBlockOffset = reader.readInt64();

        // Load Textures
        if (infoPtrArrayOffset !== 0) {
            SeekTask.run(reader, infoPtrArrayOffset, 'begin', () => {
                file.textures = [];
                const textureOffsets: number[] = [];
                for (let i = 0; i < textureCount; i++) {
                    textureOffsets.push(Number(reader.readInt64()));
                }

                for (const offset of textureOffsets) {
                    SeekTask.run(reader, offset, 'begin', () => {
                        const tex = new Texture();
                        tex.load(reader);
                        file.textures.push(tex);
                    });
                }
            });
        }

        // Texture Dictionary
        const textureDictOffset = Number(reader.readInt64());
        if (textureDictOffset !== 0) {
            SeekTask.run(reader, textureDictOffset, 'begin', () => {
                file.textureDict.load(reader);

                // Populate node keys from string table (loaded later or inline)
                // In BNTX, ResDict nodes have string offsets that we need to resolve
                const nodeReader = new BinaryDataReader(reader.buffer);
                nodeReader.byteOrder = reader.byteOrder;
                for (let i = 0; i < file.textureDict.nodes.length; i++) {
                    const node = file.textureDict.nodes[i];
                    // Re-read node to get key offset (IdxLeft/Right are at +4, +6. Key offset is at +8)
                    // Each node is 16 bytes.
                    const nodeOffset = textureDictOffset + 8 + (i * 16);
                    nodeReader.position = nodeOffset;
                    const keyOffset = Number(nodeReader.readInt64());
                    if (keyOffset !== 0) {
                        nodeReader.position = keyOffset;
                        const len = nodeReader.readUInt16();
                        node.key = nodeReader.readStringWithLength(len);
                    }
                }
            });
        }

        // Name
        if (fileNameOffset !== 0) {
            SeekTask.run(reader, fileNameOffset - 2, 'begin', () => {
                const len = reader.readUInt16();
                file.name = reader.readStringWithLength(len);
            });
        }

        // Relocation Table
        if (relocationTableOffset !== 0) {
            SeekTask.run(reader, relocationTableOffset, 'begin', () => {
                file.relocationTable.load(reader);
            });
        }

        // String Table — position after texture info pointer array, aligned
        reader.position = infoPtrArrayOffset + (textureCount * 8);
        reader.position = (reader.position + 7) & ~7;

        try {
            file.stringTable.load(reader);
        } catch (e) {
            // String table is sometimes missing or at different offset, failing silently like some parsers
            console.warn("Could not load string table:", e);
        }
    }

    private setVersionInfo(version: number): void {
        this.file.versionMajor = version >>> 24;
        this.file.versionMajor2 = (version >>> 16) & 0xFF;
        this.file.versionMinor = (version >>> 8) & 0xFF;
        this.file.versionMinor2 = version & 0xFF;
    }
}
